lib/elixir/lib/calendar/time.ex

```diff
@@ -321,6 +321,42 @@ defmodule Time do
   end
 
   @doc """
+  Adds the `number` of `unit`s to the given `time`.
+
+  This function accepts the `number` measured according to `Calendar.ISO`.
+  The time is returned in the same calendar as it was given in.
+
+  Note the result value represents time of day,
+  meaning that it is cyclic and will never go over 24 hours.
+
+  ## Examples
+
+      iex> Time.add(~T[10:00:00], 27000)
+      ~T[17:30:00.000000]
+      iex> Time.add(~T[11:00:00.005], 2400)
+      ~T[11:40:00.005000]
+      iex> Time.add(~T[00:00:00], 86399999, :millisecond)
+      ~T[23:59:59.999000]
+      iex> Time.add(~T[17:10:05], 86400)
+      ~T[17:10:05.000000]
+      iex> Time.add(~T[23:00:00], -60)
+      ~T[22:59:00.000000]
+
+  """
+  @spec add(Calendar.time, integer, System.time_unit) :: t
+  def add(%{calendar: calendar} = time, number, unit \\ :second) when is_integer(number) do
```

Comment:
I think doing the operation on ISO and then doing the conversion without the days is going to lead to loss of information. Imagine a calendar where days are 25 hours long. Now imagine we have `{2, {2,24}}` (i.e. we have 2 days and 2 hours in the ISO calendar). If we only convert the time part, we would have 2.08h in the other calendar but it should be 4h.

We should use the exact same code we use for `NaiveDateTime.add`. Something like this:

```elixir
    ppd = System.convert_time_unit(86400, :second, unit)
    {_, _, _, hour, minute, second, microsecond} =
      {0, to_day_fraction(time)}
      |> Calendar.ISO.add_day_fraction_to_iso_days(number, ppd)
      |> calendar.naive_datetime_from_iso_days()
    %Time{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}
```

Reply:
Actually, I believe we assume all calendars to consider a day to be the same thing, except they may split the day in different units. So my code above is irrelevant and the assumptions done by your code are correct.

Reply:
I would just avoid converting to microsecond if we don't need to and also avoid the calendar conversion I mentioned below, something like:

```
    iso_days = {0, to_day_fraction(time)}
    total = Calendar.ISO.iso_days_to_unit(iso_days, unit) + number
    iso_ppd = System.convert_time_unit(86400, :second, unit)
    parts = Integer.mod(total, iso_ppd)

    {hour, minute, second, microsecond} = calendar.time_from_day_fraction({parts, iso_ppd})
    %Time{hour: hour, minute: minute, second: second, microsecond: microsecond, calendar: calendar}  
```

Reply:
Avoiding microsecond leads to precision loss:
```elixir
Time.add(~T[11:00:00.005], 2400)
#=> ~T[11:40:00.000000]
```